sc2ts/utils.py

```diff
@@ -558,6 +559,107 @@ def get_imputed_pango(u):
 
         return pd.DataFrame(recombinants)
 
+    def export_recombination_node_breakpoints(self):
+        """
+        Make a dataframe with one row per recombination node breakpoint
+        Return info about the path to the mrca of left and right parents when travelling
+        up the tree to the left of the the breakpoint and to the right of the breakpoint.
+        
+        You can only list the breakpoints in recombination nodes that have 2 parents by
+        doing e.g. df.drop_duplicates('node', keep=False)
+        
+        """
+        # Note that it is more efficient to find the MRCA nodes in a single batch
+        # for all breakpoints, reusing the trees, rather than getting a new tree
+        # for each recombination node
+
+        # Sort recombinants by breakpoint location
+        forward_breakpoints = collections.defaultdict(list)
+        meta = self.nodes_metadata
+        for u in self.recombinants:
+            md = meta[u]
+            row = md["match_info"]
+            assert len(row) == 2
+            assert row[0]["strain"] == row[1]["strain"]
+            matches = {}
+            for record in row:
+                matches[record["direction"]] = Match(
+                    breakpoints=record["breakpoints"],
+                    parents=record["parents"],
+                    mutations=record["mutations"],
+                )
+            rec = Recombinant(row[0]["strain"], matches, node=u)
+            breaks = rec.matches["forward"].breakpoints[1:-1]
+            for i in range(len(breaks)):
+                forward_breakpoints[breaks[i]].append(
+                    (u, rec.is_hmm_consistent(), rec.matches["forward"].parents[i:i+2])
+                )
+        sorted_breakpoints = sorted(forward_breakpoints.keys())
+        tree_a_indexes = np.searchsorted(
```

Comment:
Oh, good spot, I meant `tree.seek_index()`. I'll change that.

Re traversing with 2 trees, it's easier to do it like this because (a) the code was originally written like that anyway (b) it generalises to positions that are not at explicit breakpoints (when the trees are the same`) and (c) I'd need to do multiple sets of parents for each breakpoint anyway, so I'd either need to keep going backwards and forwards between adjacent trees for each set of parents, or keep multiple lists of path_left and path_right, which seemed to make the code more complicated.

Reply:
OK, just fix it and I can go back later to test

Reply:
Done, thanks. And thanks for spotting the `seek_index` thing. My bad.